Add TANE-based algorithm for discovering pFDs

iliya-b opened the pull request:
Add TANE-based algorithm PFDTane for discovering probabilistic functional dependencies using per-value and per-tuple (also known as g3) error measure.

vs9h reviewed (changes requested):
I added a code review, but I also have a few more comments:
* need to fix commits:
  - make commit names more descriptive (`Fix formatting` and `Fix variable type` are non-informative)
  - `Add algorithm for finding probabilistic FDs` -> `Add algorithm for mining PFDs`
  - this is a rather bad practice to add changes, then delete them within one PR (this applies to a commit `Move error calculation method to ProbaTane class`) - it seems better to squash it with last commit
* Why are there no tests of the added algorithm in this pull request?

There is also a more important topic to discuss - about architecture:
1) How correct is it that we use the same class `model::FD` for fds and pfds?
2) How correct is it to inherit from the `TANE` class?

And i think, that: 1) acceptable 2) rather no than yes
That is, it seems to me that this (inheritance) is generally incorrect from an architectural point of view (the algorithm for mining `PFD`s is not an algorithm for mining `FD`s)

iliya-b commented:
> There is also a more important topic to discuss - about architecture:
> 
> 1. How correct is it that we use the same class `model::FD` for fds and pfds?
> 2. How correct is it to inherit from the `TANE` class?
> 
> And i think, that: 1) acceptable 2) rather no than yes That is, it seems to me that this (inheritance) is generally incorrect from an architectural point of view (the algorithm for mining `PFD`s is not an algorithm for mining `FD`s)

Regarding 2)
I agree that the algorithm for mining `PFD`s is not an algorithm for mining `FD`s.
To avoid such logic, I'd write a superclass TaneBasedAlgorithm, and make its direct children Tane and PFDTane just override the error calculation methods. What do you think?

vs9h commented:
> To avoid such logic, I'd write a superclass TaneBasedAlgorithm, and make its direct children Tane and PFDTane just override the error calculation methods. What do you think?

The solution is adequate and further I will describe in more detail the motivation, solution and what you need to pay attention to.

_Why can we reuse the `FD` class and `FDAlgorithm` to implement `PFD`:_

Inheritance means the "is" relationship. For example, `AFD` is `FD`. We say that an algorithm for mining `AFD`s is an algorithm for mining exact `FD`s, because we use them in the same way as exact dependencies. In this case, you can easily draw an analogy with the `PFD`, because the only difference is how the "error" is calculated. Therefore we can use inheritance from `FDAlgorithm` (to be more precise, from `PliBasedFDAlgorithm`)

_Solution:_

I will describe in more detail the solution that seems correct to me:
1) In the file `tane.h` we create a `TaneCommon` class (as suggested above, a class that has two virtual methods). At the same time, this class should be placed in the namespace (for example, `tane` or `details`) so as not to pollute the `algos` namespace.
2) In the same file, under this class, but already in namespace `algos`, add two implementations: for `FD` and for `PFD`.
3) Regarding the name for `PFD`-based class, I am absolutely sure that the name `PFDTane` is adequate. But for `FD` I can't be sure - I believe that it would be more correct to use the generally accepted name `Tane`. But I'm not exactly sure and would like to hear from @chernishev.

_Note:_

Using a virtual method can significantly degrade the running time of the algorithm. Therefore, you will also need to take measurements (for this you can use heavy datasets, see tests). If the time doesn't get worse, then everything is fine. Otherwise we will be forced to get rid of virtual methods.

iliya-b commented on src/core/algorithms/fd/probatane/probatane.h:
```diff
@@ -0,0 +1,21 @@
+
+namespace algos {
+
+class ProbaTane : public Tane {
```
No, probatane is not a well-name name, I think PFDTane suits better too

iliya-b commented:
I have fixed the issues, added a simple test (taking test_fd_verifier.cpp as example), add cleaned the commits history.
Regarding generalizing Tane and PFDTane, I think I got your point completely, so I can work further on changing the inheritance that way. But given the possible perfomance degrade because of virtual methods, I will have to  either thouroughly check the perfomance or rewrite the PFDTane without inheritance at all (copy-pasting the TANE algorithm). I am not sure what one is the optimal option.

While testing, I noticed that my implementation fails on specific cases because the algorithm depends on the specific sorting of the clusters in PositionListIndex.
It can be fixed in one line here https://github.com/Mstrutov/Desbordante/blob/4978ff73d508ae11f21b53525bbf5f12f9b8bae5/src/core/model/table/position_list_index.cpp#L201 replacing SortClusters  with 
`sort(new_index.begin(), new_index.end(), [&probing_table] (std::vector<int> const& a, std::vector<int> const& b) { 
 return (*probing_table)[a[0]] < (*probing_table)[b[0]]; 
});`
But, again, it can have effects on perfomance and it's required to check if it affects other algorithms which uses PLI (all the tests pass though).
Given this unfinished work, I've marked PR as draft and now going to work on these two problems.

vs9h commented:
> 3\. But for `FD` I can't be sure - I believe that it would be more correct to use the generally accepted name `Tane`. But I'm not exactly sure and would like to hear from @chernishev.

I clarified the question about the name with @chernishev and we will leave the name `Tane` for the usual `FD`-based algorithm.

iliya-b commented:
@vs9h I have updated PR. I have rewritten PFDTane class without changing the original Tane implementation (by means of copying some methods of Tane), so we can be sure that original Tane does not suffer perfomance loss.
This time I did not pursue to generalize the algorithms as I initially did, but I am thinking on doing that in future PRs taking in account the solution we came to in previous discussion.

vs9h commented on examples/comparison_pfd_vs_afd.py:
```diff
@@ -0,0 +1,32 @@
+pfds = make_set(get_pfds())
+afds = make_set(get_afds())
+
+print("pFDs \ AFDs =", pfds - afds)
```
why you write here "pFDs", not "PFDs"? I believe that it is better to use "PFDs", similar to `AFDs`

iliya-b replied:
it's from the paper http://i.stanford.edu/~anishds/publications/webdb09/webdb09.pdf

vs9h commented on src/core/algorithms/fd/pfdtane/pfdtane.h:
```diff
@@ -0,0 +1,50 @@
+    unsigned long long ExecuteInternal() override final;
+
+public:
+    config::ErrorType max_fd_error_;
+    config::ErrorType max_ucc_error_;
+    config::MaxLhsType max_lhs_;
+    ErrorMeasure error_measure_ = ErrorMeasure::_values()[0];
+
+    int count_of_fd_ = 0;
+    int count_of_ucc_ = 0;
+    long apriori_millis_ = 0;
```
Why these fields are public?

vs9h reviewed (changes requested):
I left comments, please try to improve the quality of the code if possible.

Especially pay attention to the huge `ExecuteInternal` method. I understand that this is most likely a copy-paste from Tane, but it seems to me that it needs to be carefully refactored. Perhaps we should still stick with the idea of creating a general class for Tane-based algorithms.

I also don't like tests because we expose methods that shouldn't be public. In general, I rather like the idea of testing the error calculation functions themselves than not. Surely it is possible to make these functions leak and somehow inherit from the algorithm class. Try to google something on this topic, or see if there is anything similar in the desbordante.

I would also suggest using some tools that can help you find errors to make your life easier. For example, `cppcheck` and `cpplint`. I'm sure they will help you avoid a number of errors that you usually don't want to look for manually. For example, inclusion of extra headers or non-inclusion.

iliya-b commented on src/core/algorithms/fd/pfdtane/pfdtane.cpp:
```diff
@@ -0,0 +1,397 @@
+        vertex->SetPositionListIndex(column_data.GetPositionListIndex());
+
+        // check FDs: 0->A
+        double fd_error = (this->*calculate_fd_zero_ary_error)(&column_data);
```
Now I have just one type of calculation methods, so we no longer need such dynamic invocation

iliya-b commented on src/core/algorithms/fd/pfdtane/pfdtane.h:
```diff
@@ -0,0 +1,50 @@
+
+    PFDTane();
+
+    model::PositionListIndex* GetColumnIndex(unsigned int column);  // for test
```
I removed this method

iliya-b commented:
@vs9h I've fixed all issues. 
1) Regarding ExecuteInternal method, I have decomposed it slightly to increase readability. 
2) For the purposes of testing I've made the functions of error calculation static, and now we can test them without creating the PFDTane algo instance.

Since last PR, I have also refactored and changed examples and made some minor changes to cli in order to check required parametersa

vs9h commented on src/tests/test_pfdtane.cpp:
```diff
@@ -0,0 +1,97 @@
+        : params({{onam::kCsvConfig, table_config},
+                  {onam::kError, error},
+                  {onam::kErrorMeasure, error_measure},
+                  {onam::kEqualNulls, true}}),
```
this option already has the same default value, maybe we should omit it here?

vs9h reviewed (changes requested):
I left some new comments.
And I would also like to return to the old question - regarding the duplication of the Tane code. You said:
> This time I did not pursue to generalize the algorithms as I initially did, but I am thinking on doing that in future PRs taking into account the solution we came to in previous discussion.
I would like to clarify what you mean by future here:
1) someone will fix this in the future
2) you will correct this in the future
In the first case, we know that this may never occur, or may occur after a long time.
In the second case, we can be sure that this will be fixed soon, because we simply split it into two PRs.

If this is the second case, then it is acceptable. If the first, then we should fix it right away.

iliya-b commented on src/core/algorithms/fd/pfdtane/pfdtane.h:
```diff
@@ -0,0 +1,41 @@
+#include "model/table/relation_data.h"
+
+namespace algos {
+using boost::dynamic_bitset;
```
in pfdtane.cpp I have `using Cluster = model::PositionListIndex::Cluster;` just to have a shorthand version of Cluster. Is that ok?

vs9h replied:
Yes, I will write the problem in more detail: when we add such using in a header, it will also be in all headers that include this header. We definitely don't want this. 
You can use such usings in source files.

iliya-b commented:
@vs9h I've fixed these new issues.
(2) I am going to generalize the algorithms myself in future PR.

vs9h commented on cli/cli.py:
```diff
@@ -337,6 +357,12 @@ def check_error_option_presence(task: str | None, error: str | None) -> None:
         sys.exit(1)
 
 
+def check_error_measure_option_presence(task: str | None, error_measure: str | None) -> None:
```
It seems, that we also should update function `check_error_option_presence`:
```py
if task in (Task.afd, Task.pfd) and error is None:
        click.echo(f"ERROR: Missing option '{ERROR}'.")
        sys.exit(1)
```
Because if we want find pfds, then we should also set error threshold.

vs9h reviewed (commented):
There are a few small comments before approving.

More important: break commit into logical parts (implementation of the algorithm, adding bindings, adding tests, adding examples). 
Note: do not use too abstract names (for example, not adding tests, but adding tests for PFDTane)

vs9h commented:
Also note that the PR #377 has just been merged with `clang-tidy` rules for naming check. Look how the file with descriptions of options and the variable names of option objects have changed.

iliya-b commented:
@vs9h I fixed the issues and I think PR is ready
